In [ ]:
import pandas as pd

In [ ]:
from google.colab import files
uploaded=files.upload()

Saving adult.csv to adult (1).csv


In [ ]:
columns=['Age','Workclass','Fnlwgt','Education','Education_num','Marital_status','Occupation','Relationship','Race','Sex','Capital_gain','Capital_loss','Hours_per_week','Native_country','Income']
df=pd.read_csv('adult.csv',header=None,na_values=[" ?"],names=columns)

In [ ]:
df.head()

Age          Workclass  Fnlwgt   Education  Education_num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        Marital_status          Occupation    Relationship    Race      Sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   Capital_gain  Capital_loss  Hours_per_week  Native_country  Income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [ ]:
df.isnull().sum()

Age                  0
Workclass         1836
Fnlwgt               0
Education            0
Education_num        0
Marital_status       0
Occupation        1843
Relationship         0
Race                 0
Sex                  0
Capital_gain         0
Capital_loss         0
Hours_per_week       0
Native_country     583
Income               0
dtype: int64

In [ ]:
my_df = df.dropna()

In [ ]:
my_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30162 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Age             30162 non-null  int64 
 1   Workclass       30162 non-null  object
 2   Fnlwgt          30162 non-null  int64 
 3   Education       30162 non-null  object
 4   Education_num   30162 non-null  int64 
 5   Marital_status  30162 non-null  object
 6   Occupation      30162 non-null  object
 7   Relationship    30162 non-null  object
 8   Race            30162 non-null  object
 9   Sex             30162 non-null  object
 10  Capital_gain    30162 non-null  int64 
 11  Capital_loss    30162 non-null  int64 
 12  Hours_per_week  30162 non-null  int64 
 13  Native_country  30162 non-null  object
 14  Income          30162 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [ ]:
my_df.isnull().sum()

Age               0
Workclass         0
Fnlwgt            0
Education         0
Education_num     0
Marital_status    0
Occupation        0
Relationship      0
Race              0
Sex               0
Capital_gain      0
Capital_loss      0
Hours_per_week    0
Native_country    0
Income            0
dtype: int64

In [ ]:
from sklearn.svm import SVR
from sklearn.preprocessing import LabelEncoder
my_df = my_df.apply(LabelEncoder().fit_transform)
my_df.head()

Age  Workclass  Fnlwgt  Education  Education_num  Marital_status  \
0   22          5    2491          9             12               4   
1   33          4    2727          9             12               2   
2   21          2   13188         11              8               0   
3   36          2   14354          1              6               2   
4   11          2   18120          9             12               2   

   Occupation  Relationship  Race  Sex  Capital_gain  Capital_loss  \
0           0             1     4    1            24             0   
1           3             0     4    1             0             0   
2           5             1     4    1             0             0   
3           5             0     2    1             0             0   
4           9             5     2    0             0             0   

   Hours_per_week  Native_country  Income  
0              39              38       0  
1              12              38       0  
2              39              38       0  
3              39              38       0  
4              39               4       0

In [ ]:
drop_elements = ['Education', 'Native_country', 'Income']
y = my_df["Income"]
x = my_df.drop(drop_elements, axis=1)

In [ ]:
from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=12)


In [ ]:
from sklearn.metrics import precision_score, accuracy_score
from sklearn.metrics import recall_score, f1_score


# Decision Tree
---



In [ ]:
import numpy as np
from sklearn.tree import DecisionTreeClassifier
tree=DecisionTreeClassifier(criterion='gini',max_depth=10,min_samples_split=25)
tree.fit(x_train,y_train)

DecisionTreeClassifier(max_depth=10, min_samples_split=25)

In [ ]:
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score

print('Tree-training score: ',tree.score(x_train,y_train))
print('Tree-testing score: ',tree.score(x_test,y_test))

Tree-training score:  0.8658875212400017
Tree-testing score:  0.8461793469252444


In [ ]:
dt_predict = tree.predict(x_test)
cm_dt = confusion_matrix(y_test,dt_predict)
cr_dt = classification_report(y_test,dt_predict)
print('Confusion matrix\n',cm_dt)
print('\n')
print('Classifiation report\n',cr_dt)

print('\nPrecision: ', precision_score(y_test, dt_predict))
print('Recall: ', recall_score(y_test, dt_predict))
print('F1 Score: ', f1_score(y_test, dt_predict))
print("Accuracy Rate: ", accuracy_score(y_test,dt_predict))

Confusion matrix
 [[4176  390]
 [ 538  929]]


Classifiation report
               precision    recall  f1-score   support

           0       0.89      0.91      0.90      4566
           1       0.70      0.63      0.67      1467

    accuracy                           0.85      6033
   macro avg       0.80      0.77      0.78      6033
weighted avg       0.84      0.85      0.84      6033


Precision:  0.7043214556482184
Recall:  0.6332651670074982
F1 Score:  0.6669059583632447
Accuracy Rate:  0.8461793469252444


# KNN Classifier

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

In [ ]:
knn = KNeighborsClassifier(n_neighbors = 8)
knn.fit(x_train,y_train)

KNeighborsClassifier(n_neighbors=8)

In [ ]:
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score

print('KNN-training score: ',knn.score(x_train,y_train))
print('KNN-testing score: ',knn.score(x_test,y_test))

KNN-training score:  0.8138339757138713
KNN-testing score:  0.7898226421349246


In [ ]:
knn_predict = knn.predict(x_test)

In [ ]:
cm_knn = confusion_matrix(y_test,knn_predict)
cr_knn = classification_report(y_test,knn_predict)
print('Confusion Matrix\n',cm_knn)
print('\n')
print('Classification Report\n',cr_knn)

print('\nPrecision: ', precision_score(y_test, knn_predict))
print('Recall: ', recall_score(y_test, knn_predict))
print('F1 Score: ', f1_score(y_test, knn_predict))
print("Accuracy Rate: ", accuracy_score(y_test,knn_predict))

Confusion Matrix
 [[4400  166]
 [1102  365]]


Classification Report
               precision    recall  f1-score   support

           0       0.80      0.96      0.87      4566
           1       0.69      0.25      0.37      1467

    accuracy                           0.79      6033
   macro avg       0.74      0.61      0.62      6033
weighted avg       0.77      0.79      0.75      6033


Precision:  0.687382297551789
Recall:  0.24880708929788684
F1 Score:  0.36536536536536535
Accuracy Rate:  0.7898226421349246


# SVM 

In [ ]:
from sklearn.svm import SVC

In [ ]:
classifier = SVC(kernel = 'linear', random_state = 0)
classifier.fit(x_train, y_train)
y_pred = classifier.predict(x_test)

In [ ]:
cm_svm = confusion_matrix(y_test,y_pred)
cr_svm = classification_report(y_test,y_pred)

print('Confusion Matrix\n',cm_svm)
print('\n')
print('Classification Report\n',cr_svm)

print('\nPrecision: ', precision_score(y_test, y_pred))
print('Recall: ', recall_score(y_test, y_pred))
print('F1 Score: ', f1_score(y_test, y_pred))
print("Accuracy Rate: ", accuracy_score(y_test,y_pred))

Confusion Matrix
 [[4266  300]
 [ 866  601]]


Classification Report
               precision    recall  f1-score   support

           0       0.83      0.93      0.88      4566
           1       0.67      0.41      0.51      1467

    accuracy                           0.81      6033
   macro avg       0.75      0.67      0.69      6033
weighted avg       0.79      0.81      0.79      6033


Precision:  0.6670366259711432
Recall:  0.4096796182685753
F1 Score:  0.5076013513513514
Accuracy Rate:  0.8067296535720205


# Random Forest

In [ ]:
#RFF
from sklearn.ensemble import RandomForestClassifier
rff=RandomForestClassifier(n_estimators=10)


In [ ]:

rff.fit(x_train,y_train)
rff_pred=rff.predict(x_test)


In [ ]:
from sklearn import metrics
from sklearn.metrics import confusion_matrix,classification_report

cm_rff = confusion_matrix(y_test,rff_pred)
cr_rff = classification_report(y_test,rff_pred)

print('Confusion Matrix\n',cm_rff)
print('\n')
print('Classification Report\n',cr_rff)

print('\nPrecision: ', precision_score(y_test, rff_pred))
print('Recall: ', recall_score(y_test, rff_pred))
print('F1 Score: ', f1_score(y_test, rff_pred))
print("Accuracy Rate: ", accuracy_score(y_test,rff_pred))

Confusion Matrix
 [[4272  294]
 [ 635  832]]


Classification Report
               precision    recall  f1-score   support

           0       0.87      0.94      0.90      4566
           1       0.74      0.57      0.64      1467

    accuracy                           0.85      6033
   macro avg       0.80      0.75      0.77      6033
weighted avg       0.84      0.85      0.84      6033


Precision:  0.738898756660746
Recall:  0.567143830947512
F1 Score:  0.6417277284998072
Accuracy Rate:  0.8460135919111553


# Logistic Regression

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline

In [ ]:
lr = make_pipeline(StandardScaler(), LogisticRegression())
lr.fit(x_train,y_train)


Pipeline(steps=[('standardscaler', StandardScaler()),
                ('logisticregression', LogisticRegression())])

In [ ]:
print('Training score: ',lr.score(x_train,y_train))
print('Testing Score: ',lr.score(x_test,y_test))

Training score:  0.8203406689046375
Testing Score:  0.8143543842201226


In [ ]:
lr_predict = lr.predict(x_test)
cm_lr = confusion_matrix(y_test,lr_predict)
cr_lr = classification_report(y_test,lr_predict)

print('Confusion Matrix\n',cm_lr)
print('\n')
print('Classification Report\n',cr_lr)

print('\nPrecision: ', precision_score(y_test,lr_predict))
print('Recall: ', recall_score(y_test, lr_predict))
print('F1 Score: ', f1_score(y_test, lr_predict))
print("Accuracy Rate: ", accuracy_score(y_test,lr_predict))

Confusion Matrix
 [[4263  303]
 [ 817  650]]


Classification Report
               precision    recall  f1-score   support

           0       0.84      0.93      0.88      4566
           1       0.68      0.44      0.54      1467

    accuracy                           0.81      6033
   macro avg       0.76      0.69      0.71      6033
weighted avg       0.80      0.81      0.80      6033


Precision:  0.6820566631689402
Recall:  0.4430811179277437
F1 Score:  0.5371900826446281
Accuracy Rate:  0.8143543842201226


In [ ]:
# Misclassification percentage
ms_dt = 1-accuracy_score(y_test,dt_predict)
ms_rf = 1-accuracy_score(y_test,rff_pred)
ms_lr = 1-accuracy_score(y_test,lr_predict)
ms_knn = 1-accuracy_score(y_test,knn_predict)
ms_svm = 1-accuracy_score(y_test,y_pred)
print('Misclassification rate of Decision Tree: {} , Random Forest: {} , Logistic Regression: {} , KNN: {} , SVM: {} '.format(ms_dt,ms_rf,ms_lr,ms_knn,ms_svm))

Misclassification rate of Decision Tree: 0.15382065307475556 , Random Forest: 0.15398640808884467 , Logistic Regression: 0.18564561577987737 , KNN: 0.2101773578650754 , SVM: 0.19327034642797947 


# Misclassification Percentage

Misclassification rate = 1 - Accuracy

1. Misclassification of Decision Tree = 1- 0.8460 ~= 0.154 ~= 15.4%
2. Misclassification of Random Forest = 1- 0.8480 ~= 0.152 ~= 15.2%
3. Misclassification of Logistic Regression = 1- 0.8143 ~= 0.18.57 ~= 18.57%
4. Misclassification of KNN  = 1- 0.7898 ~= 0.2102 ~= 21.02%
5. Misclassification of SVM Classifier = 1- 0.8067 ~= 0.1933 ~= 19.33%


# Report
# Accuracy Based best Model is Random Forest Classifer with 84% Accuracy